File: app/src/main/java/com/ou/demo/opengl/YuvPlayer.java
package com.ou.demo.opengl;

import android.content.Context;
import android.opengl.GLSurfaceView;
import android.util.AttributeSet;
import android.util.Log;
import android.view.SurfaceHolder;

import com.ou.demo.util.Util;

import javax.microedition.khronos.egl.EGLConfig;
import javax.microedition.khronos.opengles.GL10;

public class YuvPlayer extends GLSurfaceView implements Runnable, SurfaceHolder.Callback, GLSurfaceView.Renderer {

    public YuvPlayer(Context context, AttributeSet attrs) {
        super(context, attrs);
        setRenderer(this);
        setRenderMode(RENDERMODE_CONTINUOUSLY);
        Util.INSTANCE.log("init surface = " + getHolder().getSurface());
    }

    @Override
    public void run() {
        drawTriangle(getHolder().getSurface());
    }

    @Override
    public void surfaceCreated(SurfaceHolder holder) {
        Log.d("okunu","surfaceCreated ---- ");
        Util.INSTANCE.log("surface = " + getHolder().getSurface());
        //在Surface创建的时候启动绘任务
        new Thread(this).start();
    }

    @Override
    public void surfaceDestroyed(SurfaceHolder holder) {

    }

    @Override
    public void surfaceChanged(SurfaceHolder holder, int format, int w, int h) {

    }

    @Override
    public void onSurfaceCreated(GL10 gl, EGLConfig config) {
        Log.d("okunu","onSurfaceCreated");
    }

    @Override
    public void onSurfaceChanged(GL10 gl, int width, int height) {
        Log.d("okunu","onSurfaceChanged");
    }

    @Override
    public void onDrawFrame(GL10 gl) {
        Log.d("okunu","onDrawFrame");
    }

    public native void drawTriangle(Object surface);
}

File: app/src/main/java/com/ou/demo/opengl/YuvPlayer.kt
package com.ou.demo.opengl

import android.content.Context
import android.content.res.AssetManager
import com.ou.demo.util.Util.log
import android.opengl.GLSurfaceView
import android.util.AttributeSet
import android.util.Log
import android.view.SurfaceHolder
import javax.microedition.khronos.egl.EGLConfig
import javax.microedition.khronos.opengles.GL10

class YuvPlayer(context: Context?, attrs: AttributeSet?) : GLSurfaceView(context, attrs), Runnable,
    SurfaceHolder.Callback, GLSurfaceView.Renderer {
    init {
        setRenderer(this)
        renderMode = RENDERMODE_CONTINUOUSLY
        log("init surface = " + holder.surface)
    }

    override fun run() {
//        drawTriangle(holder.surface)
        drawYuv(holder.surface, context.assets)
    }

    override fun surfaceCreated(holder: SurfaceHolder) {
        Log.d("okunu", "surfaceCreated ---- ")
        log("surface = " + getHolder().surface)
        //在Surface创建的时候启动绘任务
        Thread(this).start()
    }

    override fun surfaceDestroyed(holder: SurfaceHolder) {}
    override fun surfaceChanged(holder: SurfaceHolder, format: Int, w: Int, h: Int) {}
    override fun onSurfaceCreated(gl: GL10, config: EGLConfig) {
        Log.d("okunu", "onSurfaceCreated")
    }

    override fun onSurfaceChanged(gl: GL10, width: Int, height: Int) {
        Log.d("okunu", "onSurfaceChanged")
    }

    override fun onDrawFrame(gl: GL10) {
        Log.d("okunu", "onDrawFrame")
    }

    external fun drawTriangle(surface: Any?)
    external fun drawYuv(surface: Any?, manager: AssetManager)
}